namerd/storage/consul/src/main/scala/io/buoyant/namerd/storage/consul/ConsulDtabStore.scala

```diff
@@ -0,0 +1,102 @@
+import com.twitter.io.Buf
+import com.twitter.util._
+import io.buoyant.consul.NotFound
+import io.buoyant.consul.v1._
+
+import io.buoyant.namerd.DtabStore.{DtabNamespaceAlreadyExistsException, DtabNamespaceDoesNotExistException}
+
+class ConsulDtabStore(api: KvApi, root: Path) extends DtabStore {
+  import DtabStore.{DtabVersionMismatchException, Version}
+
+  override def list(): Activity[Set[Ns]] = {
+    def namespace(key: String): Ns = key.stripPrefix("/").stripSuffix("/").substring(root.show.length)
+
+    val run = Var.async[Activity.State[Set[Ns]]](Activity.Pending) { updates =>
+      @volatile var running = true
+
+      def cycle(index: Option[String]): Future[Unit] =
+        if (running)
+          api.list(s"${root.show}/", blockingIndex = index) transform {
+            case Return(result) =>
+              val namespaces = result.value.map(namespace).toSet
+              updates() = Activity.Ok(namespaces)
+              cycle(result.index)
+            case Throw(e) =>
+              updates() = Activity.Failed(e)
+              cycle(None)
+          }
+        else
+          Future.Unit
+      val pending = cycle(None)
+
+      Closable make { _ =>
+        running = false
+        pending.raise(new FutureCancelledException)
+        Future.Unit
+      }
+    }
+
+    Activity(run)
+  }
+
+  def create(ns: Ns, dtab: Dtab): Future[Unit] =
+    api.put(s"${root.show}/$ns", dtab.show, cas = Some("0")) flatMap { result =>
+      if (result) Future.Done else Future.exception(new DtabNamespaceAlreadyExistsException(ns))
+    }
+
+  def delete(ns: Ns): Future[Unit] = {
+    val key = s"${root.show}/$ns"
+    api.get(key) transform {
+      case Return(_) => api.delete(key).unit
+      case Throw(e: NotFound) => Future.exception(new DtabNamespaceDoesNotExistException(ns))
+      case Throw(e) => Future.exception(e)
+    }
+  }
+
+  def update(ns: Ns, dtab: Dtab, version: Version): Future[Unit] = {
+    val Buf.Utf8(vstr) = version
+    api.put(s"${root.show}/$ns", dtab.show, cas = Some(vstr)) flatMap { result =>
+      if (result) Future.Done else Future.exception(new DtabVersionMismatchException)
+    }
+  }
+
+  def put(ns: Ns, dtab: Dtab): Future[Unit] =
+    api.put(s"${root.show}/$ns", dtab.show).unit
+
+  def observe(ns: Ns): Activity[Option[VersionedDtab]] = {
+    val key = s"${root.show}/$ns"
+    val run = Var.async[Activity.State[Option[VersionedDtab]]](Activity.Pending) { updates =>
+      @volatile var running = true
+
+      def cycle(index: Option[String]): Future[Unit] =
+        if (running)
+          api.get(key, blockingIndex = index) transform {
+            case Return(result) =>
+              val version = Buf.Utf8(result.index.get)
+              val dtab = Dtab.read(result.value)
+              updates() = Activity.Ok(Some(VersionedDtab(dtab, version)))
+              cycle(result.index)
+            case Throw(e: NotFound) =>
+              updates() = Activity.Ok(None)
+              cycle(None)
```

Comment:
in the case of a NotFound, we still want to extract the `X-Consul-Index` and set the index, otherwise we'll just send keep sending requests here.